mat64/cholesky.go

```diff
@@ -7,84 +7,109 @@ package mat64
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/internal/asm"
+)
 
-// CholeskyL returns the left Cholesky decomposition of the matrix a and whether
-// the matrix is symmetric and positive definite. The returned matrix l is a lower
-// triangular matrix such that a = l.l'.
-func Cholesky(a *Dense) CholeskyFactor {
-	// Initialize.
-	m, n := a.Dims()
-	spd := m == n
-	l := NewDense(n, n, nil)
+// Cholesky calculates the Cholesky decomposition of the matrix A and returns
+// whether the matrix is positive definite. The returned matrix is either a
+// lower triangular matrix such that A = L * L^T or an upper triangular matrix
+// such that A = U^T * U depending on the upper parameter.
+func (t *Triangular) Cholesky(a *SymDense, upper bool) (ok bool) {
+	_, n := a.Dims()
+	if t.isZero() {
+		t.mat = blas64.Triangular{
+			N:      n,
+			Stride: n,
+			Diag:   blas.NonUnit,
+			Data:   use(t.mat.Data, n*n),
+		}
+	} else if n != t.mat.N {
+		panic(ErrShape)
+	}
+	mat := t.mat.Data
 
-	// Main loop.
-	for j := 0; j < n; j++ {
-		lRowj := l.RawRowView(j)
-		var d float64
-		for k := 0; k < j; k++ {
-			var s float64
-			for i, v := range l.RawRowView(k)[:k] {
-				s += v * lRowj[i]
+	if upper {
+		t.mat.Uplo = blas.Upper
+		for j := 0; j < n; j++ {
+			var d float64
+			for k := 0; k < j; k++ {
+				s := asm.DdotInc(
+					mat, mat,
+					uintptr(k),
+					uintptr(n), uintptr(n),
+					uintptr(k), uintptr(j),
+				)
+				s = (a.at(j, k) - s) / t.at(k, k)
+				t.set(k, j, s)
+				d += s * s
+			}
+			d = a.at(j, j) - d
+			if d <= 0 {
+				t.Reset()
+				return false
 			}
-			s = (a.at(j, k) - s) / l.at(k, k)
-			lRowj[k] = s
-			d += s * s
-			spd = spd && a.at(k, j) == a.at(j, k)
+			t.set(j, j, math.Sqrt(math.Max(d, 0)))
 		}
-		d = a.at(j, j) - d
-		spd = spd && d > 0
-		l.set(j, j, math.Sqrt(math.Max(d, 0)))
-		for k := j + 1; k < n; k++ {
-			l.set(j, k, 0)
+	} else {
+		t.mat.Uplo = blas.Lower
+		for j := 0; j < n; j++ {
+			var d float64
+			for k := 0; k < j; k++ {
+				// Short slice must go first.
+				s := asm.DdotUnitary(mat[j*n:], mat[k*n:])
+				s = (a.at(j, k) - s) / t.at(k, k)
+				t.set(j, k, s)
+				d += s * s
+			}
+			d = a.at(j, j) - d
+			if d <= 0 {
+				t.Reset()
+				return false
+			}
+			t.set(j, j, math.Sqrt(math.Max(d, 0)))
 		}
 	}
 
-	return CholeskyFactor{L: l, SPD: spd}
+	return true
 }
 
-// CholeskySolve returns a matrix x that solves a.x = b where a = l.l'. The matrix b must
-// have the same number of rows as a, and a must be symmetric and positive definite. The
-// matrix b is overwritten by the operation.
-func (f CholeskyFactor) Solve(b *Dense) (x *Dense) {
-	if !f.SPD {
-		panic("mat64: matrix not symmetric positive definite")
-	}
-	l := f.L
-
-	m, n := l.Dims()
+// SolveTri finds the matrix x that solves A * X = B where A = L * L^T or
+// A = U^T * U, and U or L are represented by t. The matrix A must be symmetric
+// and positive definite.
+func (m *Dense) SolveTri(t *Triangular, b Matrix) error {
```

Comment:
This should probably be SolveChol, so that SolveTri can be the direct wrapping around Dtrsm. Sorry for my lack of understanding previously.